### PR TITLE
Remove shard data files when they fail to write for snapshot

### PR DESCRIPTION
**Pull request by ywangd**
When shard level data files fail to write for snapshots, these data files become dangling and will not be referenced anywhere. Today we leave them as is. If the failure is disk full, the repository will become unusable because even the delete snapshot operation requires write a small metadata file first.

This PR adds a clean up step so that the shard data files are removed if the shard data files fail to snapshot.

Relates: #67790

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -2919,45 +2920,65 @@ private void doSnapshotShard(SnapshotShardContext context) {
...
-                );
-                snapshotStatus.moveToDone(threadPool.absoluteTimeInMillis(), shardSnapshotResult);
-                delegate.onResponse(shardSnapshotResult);
+            }, e -> {
+                try {
+                    shardContainer.deleteBlobsIgnoringIfNotExists(
+                        copyOfFilesToSnapshot.stream()
+                            .flatMap(fileInfo -> IntStream.range(0, fileInfo.numberOfParts()).mapToObj(fileInfo::partName))
+                            .iterator()
+                    );
```
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:2975]: This part is the actual change.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 2]
```diff
@@ -2919,45 +2920,65 @@ private void doSnapshotShard(SnapshotShardContext context) {
...
-                    getSegmentInfoFileCount(blobStoreIndexShardSnapshot.indexFiles())
-                );
-                snapshotStatus.moveToDone(threadPool.absoluteTimeInMillis(), shardSnapshotResult);
-                delegate.onResponse(shardSnapshotResult);
+            }, e -> {
+                try {
+                    shardContainer.deleteBlobsIgnoringIfNotExists(
+                        copyOfFilesToSnapshot.stream()
+                            .flatMap(fileInfo -> IntStream.range(0, fileInfo.numberOfParts()).mapToObj(fileInfo::partName))
+                            .iterator()
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:2974]: Nit: streams are kinda heavyweight, we can avoid them here:

```suggestion
                        Iterators.flatMap(copyOfFilesToSnapshot.iterator(), f -> Iterators.forRange(0, f.numberOfParts(), f::partName))
```

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 3]
```diff
@@ -2919,45 +2920,65 @@ private void doSnapshotShard(SnapshotShardContext context) {
                 };
             }
 
+            // filesToSnapshot will be emptied while snapshotting the file. We make a copy here for cleanup purpose in case of failure.
+            final List<FileInfo> copyOfFilesToSnapshot = List.copyOf(filesToSnapshot);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:2924]: I wonder whether it'd be nicer to make this `AtomicReference<List<FileInfo>> filesToCleanUp` which we set to `null` when we pass the point of no return. The mix of try/catch blocks and `ActionListener.wrap` calls is kinda hard to follow IMO, I'd rather we did everything with `ActionListener` utils if possible.

File: server/src/test/java/org/elasticsearch/repositories/fs/FsRepositoryTests.java [hunk 4]
```diff
@@ -192,6 +208,151 @@ public void testSnapshotAndRestore() throws IOException {
         }
     }
 
+    public void testCleanUpWhenShardDataFilesFailToWrite() throws IOException {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/repositories/fs/FsRepositoryTests.java:211]: :+1: looks very thorough.

**Comment by DaveCTurner** [comment]
> Resolves: https://github.com/elastic/elasticsearch/issues/67790

I'd rather we didn't say this _resolves_ #67790. It's probably enough in practice, but it is not a complete solution in all cases. After merging this I would like to close #67790 with a separate comment pointing to this change and describing its limitations and why we think we won't be doing any further work in this area.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 2 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:2974]: TIL

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 3 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Great suggestion. It made the diff much nicer to look at. Thanks!

**Comment by ywangd** [comment]
> I'd rather we didn't say this resolves https://github.com/elastic/elasticsearch/issues/67790.

I changed it to `Relates`.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 5]
```diff
@@ -2919,8 +2920,10 @@ private void doSnapshotShard(SnapshotShardContext context) {
                 };
             }
 
+            // filesToSnapshot will be emptied while snapshotting the file. We make a copy here for cleanup purpose in case of failure.
+            final AtomicReference<List<FileInfo>> fileToCleanUp = new AtomicReference<>(List.copyOf(filesToSnapshot));
             final ListenableFuture<Collection<Void>> allFilesUploadedListener = new ListenableFuture<>();
-            allFilesUploadedListener.addListener(context.delegateFailureAndWrap((delegate, v) -> {
+            allFilesUploadedListener.addListener(ActionListener.wrap(ignore -> {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:2926]: somewhat-unrelated nit: this is equivalent (and simpler) right?

```
            final ActionListener<Collection<Void>> allFilesUploadedListener = ActionListener.assertOnce(ActionListener.wrap(ignore -> {
```

**Review by DaveCTurner — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 5 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: I was concerned that changing it from `ListenableFuture` to a plain `ActionListener` could cause test failures due to how exception is wrapped by `ListenableFuture`. But running some tests locally seems to be fine. I have updated the PR with the suggestion and will see whether CI is happy with it.
